### ISSUE TITLE
Loss metric to use required_output_keys

### ISSUE THREAD

**Issue by vfdev-5** [pytorch/ignite#1415]
## 🚀 Feature

Currently, if we have custom metrics that require data other then `y_pred` and `y`, [we suggest](https://discuss.pytorch.org/t/how-access-inputs-in-custom-ignite-metric/91221/6) to do the following: 
```python
metrics = {
    "Accuracy": Accuracy(),
    "Loss": Loss(criterion, output_transform=lambda out_dict: (out_dict["y_pred"], out_dict["y"])),
    "CustomMetric": CustomMetric()
}

evaluator = create_supervised_evaluator(
    model, 
    metrics=metrics, 
    output_transform=lambda x, y, y_pred: {"x": x, "y": y, "y_pred": y_pred}
)
```

where `CustomMetric` is defined as 
```python
class CustomMetric(Metric):

    required_output_keys = ("y_pred", "y", "x")
```

The idea is to extend this for `Loss` metric to support `required_output_keys`. The main issue with `Loss` now is with `(prediction, target, kwargs)` optional input, where `kwargs` is a dict for extra args for criterion function.

**Comment by sdesrozis** [comment]
So, `output_transform` signature used in `Loss` initialization should be defined in a generic way in the `Loss` class ? and this signature is related to the criterion ? 

The output dict is compared to `required_output_keys` and here, we need to handle extra args defined in criterion... Not so easy because requirements are not statically known.

**Comment by vfdev-5** [comment]
@sdesrozis yes, you are right about static vs dynamic params. We can adopt two approaches:
1) if user would like to pass kwargs for the criterion, there is no way to use dict as state.output
2) in `Loss` we define `required_output_keys=("y_pred", "y", "criterion_kwargs")` and ask the user to define `state.output` as 
```python
state.output = {
  "y_pred": ...,
  "y": ...,
  "criterion_kwargs": {...}
}
```

**Comment by madhucharan** [comment]
Hello @vfdev-5 , I am new to PyTorch ignite and studied this issue for some time and understood some part of it related to loss.py but stuck with criterion. Would it be possible to explain the expectation of this issue and some information about the criterion function as well as a state? output. I am interested to work on this issue.

**Comment by vfdev-5** [comment]
@madhucharan thanks for proposing a help with this issue. Please read https://pytorch.org/ignite/metrics.html#ignite.metrics.Metric.required_output_keys about `required_output_keys`. 
Please, see our [CONTRIBUTING guidelines](https://github.com/pytorch/ignite/blob/master/CONTRIBUTING.md). You can also take a look at Quick-start and Concepts sections of our [docs](https://pytorch.org/ignite/).

**Comment by madhucharan** [comment]
In pytorch we generally use like this,
`criterion = nn.CrossEntropyLoss()`
in training
we use 
`loss = criterion(output, target)`

Is this the criterion mentioned in the first comment? or something else. Kindly clarify this. 

2. And the state.output is actually an output returned by run() function right? I would like to know why user needs to define state.output dict and how and where it is passed as an argument?

**Comment by vfdev-5** [comment]
@madhucharan well, in most of the cases a loss function in pytorch takes predictions and targets, but there are also cases where loss function can accept other params (e.g. `reduction` for functional cross entropy) and also inputs like sample x etc.
See the link in the first message (https://discuss.pytorch.org/t/how-access-inputs-in-custom-ignite-metric/91221).

The idea of this issue is to simplify Ignite's part of handling those functions.
Try to play with the code snippet from here : https://discuss.pytorch.org/t/how-access-inputs-in-custom-ignite-metric/91221/6

> And the state.output is actually an output returned by run() function right? I would like to know why user needs to define state.output dict and how and where it is passed as an argument?

Passing a dict as the output of the evaluation step function (which is written to `state.output`) is a nice way to structure the output.
In most of the cases, output is a tuple of `(y_pred, y_true)` which are used to compute metrics.

Hope this helps to understand.

**Comment by sparkingdark** [comment]
@vfdev-5 are we going to create an extended class `customMetrics` where users can use custom operations like `reset`, `update`, and `compute`.

**Comment by vfdev-5** [comment]
@sparkingdark no, the idea is to improve existing `Loss` class

**Comment by sparkingdark** [comment]
Okay

**Comment by 01-vyom** [comment]
@vfdev-5, I was studying this issue for the last two days as I wanted to contribute to PyTorch as a newbie. I looked at all the code and all the comments, and here is what I understood, we already provide a `required_output_key` argument for the `Metric` class and by that, we provide any extra params we may need while calculating custom metric. We need to extend this to the `Loss` metric, where currently the `required_output_key` is `None` ([here](https://pytorch.org/ignite/master/_modules/ignite/metrics/loss.html#Loss)).  So we can do something like this for custom loss :

```python
class CustomLoss(Loss):

    required_output_keys = ("y_pred", "y", "criterion_kwargs")

    def __init__(
        self,
        loss_fn: Callable,
        output_transform: Callable = lambda x: x,
        batch_size: Callable = len,
        device: Union[str, torch.device] = torch.device("cpu"),
    ):
        super(Loss, self).__init__(output_transform, device=device)
        self._loss_fn = loss_fn
        self._batch_size = batch_size

    def reset(self) -> None:

    def update(self, output: Sequence[Union[torch.Tensor, Dict]]) -> None:

    def compute(self) -> float:

```

Then, while calling custom loss we simply do the following:

```python
metrics = {
    "Accuracy": Accuracy(),
    "CustomLoss": CustomLoss(criterion, output_transform=lambda out_dict: (out_dict["y_pred"], out_dict["y"], out_dict["criterion_kwargs"])),
    "CustomMetric": CustomMetric(output_transform=lambda out_dict: (out_dict["y_pred"], out_dict["y"], out_dict["x"])
}

evaluator = create_supervised_evaluator(
    model, 
    metrics=metrics, 
    output_transform=lambda x, y, y_pred, criterion_kwargs, : {"x": x, "y": y, "y_pred": y_pred, "criterion_kwargs":criterion_kwargs}
)
```

Let me know if I have understood this correctly.

**Comment by vfdev-5** [comment]
@01-vyom thanks for the time on studying this issue. I agree it is not clearly stated what we would like to do here. Sorry about that.

The only thing to do here is to update current implementation of `Loss` by defining `required_output_keys = ("y_pred", "y", "criterion_kwargs")` instead of None and updating the docs saying that we can now interpret output's keys if it is a dictionary like here : https://pytorch.org/ignite/metrics.html#ignite.metrics.Metric.required_output_keys

Please, let me know if it stille unclear ?

**Comment by 01-vyom** [comment]
Yes, it is clear. I will change the argument and add a demo code as well as a statement that we accept a dictionary argument to the docs. I will make a PR soon. Thank you for explaining the issue.
